File: src/main/java/org/demoth/aworlds/server2/LocationWorkerManager.java
package org.demoth.aworlds.server2;

import org.demoth.aworlds.server2.api.LongPropertiesEnum;
import org.demoth.aworlds.server2.api.messaging.Message;
import org.demoth.aworlds.server2.api.messaging.fromServer.Positioned;
import org.demoth.aworlds.server2.api.messaging.fromServer.StateChangeData;
import org.demoth.aworlds.server2.model.Location;
import org.demoth.aworlds.server2.model.Player;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Collection;
import java.util.concurrent.ConcurrentHashMap;

import static org.demoth.aworlds.server2.api.LongPropertiesEnum.X;
import static org.demoth.aworlds.server2.api.LongPropertiesEnum.Y;

@Component
@Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
public class LocationWorkerManager {

    private static final Logger LOG = LoggerFactory.getLogger(LocationWorkerManager.class);

    private final ConcurrentHashMap<Location, Thread> instances = new ConcurrentHashMap<>();

    public void runLocation(Location location) {
        Thread worker = new Thread(() -> {
            LOG.debug("Start updating {}", location.getName());
            while (true) {
                if (location.getPlayers().isEmpty()) {
                    LOG.debug("No players in {}, exiting...", location.getName());
                    return;
                }
                Collection<Message> updates = location.updateLocation();
                // todo move to other thread
                filterUpdates(updates, location.getPlayers());
                boolean allPlayersReady = location.getPlayers().stream().noneMatch(Player::idle);
                // todo: make configurable
                long sleep = (long) (allPlayersReady ? 100 : 2000);
                try {
                    Thread.sleep(sleep);
                } catch (InterruptedException e) {
                    LOG.debug("Finishing working thread: {}", Thread.currentThread().getName());
                    // todo persist location
                    return;
                }
            }
        });
        instances.put(location, worker);
        worker.setName("Location worker: " + location.getName());
        worker.start();
    }

    public void stopLocation(Location location) {
        Thread worker = instances.get(location);
        if (worker != null)
            worker.interrupt();
    }

    /*
     * For each player filter updates and put them to the send queue.
     */
    private void filterUpdates(Collection<Message> updates, Collection<Player> players) {
        players.forEach(player -> {
            Long radius = player.getLong(LongPropertiesEnum.SIGHT_RADIUS);
            updates.stream().filter(message -> {
                        if (message instanceof StateChangeData) {
                            StateChangeData data = (StateChangeData) message;
                            return player.getId().equals(data.id);
                        }
                        if (message instanceof Positioned) {
                            Positioned appear = (Positioned) message;
                            if (Math.abs(appear.getX() - player.getLong(X)) > radius)
                                return false;
                            if (Math.abs(appear.getY() - player.getLong(Y)) > radius)
                                return false;
                            return true;
                        }
                return false;
                    }
            ).forEach(player::enqueueResponse);
        });
    }
}

File: src/main/kotlin/org/demoth/aworlds/server2/LocationWorkerManager.kt
package org.demoth.aworlds.server2

import org.demoth.aworlds.server2.api.LongPropertiesEnum
import org.demoth.aworlds.server2.api.LongPropertiesEnum.X
import org.demoth.aworlds.server2.api.LongPropertiesEnum.Y
import org.demoth.aworlds.server2.api.messaging.Message
import org.demoth.aworlds.server2.api.messaging.AppearData
import org.demoth.aworlds.server2.api.messaging.StateChangeData
import org.demoth.aworlds.server2.model.Location
import org.demoth.aworlds.server2.model.Player
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap

@Component
@Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
open class LocationWorkerManager {

    private val instances = ConcurrentHashMap<Location, Thread>()

    fun runLocation(location: Location) {
        val worker = Thread {
            LOG.debug("Start updating {}", location.name)
            while (true) {
                if (location.players.isEmpty()) {
                    LOG.debug("No players in {}, exiting...", location.name)
                    return@Thread
                }
                val updates = location.updateLocation()
                val allPlayersReady = location.players.stream().noneMatch({ it.idle() })
                // todo: make configurable
                val sleep = if (allPlayersReady) 100 else 2000
                try {
                    Thread.sleep(sleep.toLong())
                } catch (e: InterruptedException) {
                    LOG.debug("Finishing working thread: {}", Thread.currentThread().name)
                    // todo persist location
                    return@Thread
                }

            }
        }
        instances[location] = worker
        worker.name = "Location worker: " + location.name!!
        worker.start()
    }

    fun stopLocation(location: Location) {
        val worker = instances[location]
        worker?.interrupt()
    }

    /*
     * For each player filter updates and put them to the send queue.
     */
    @Deprecated("")
    private fun filterUpdates(updates: Collection<Message>, players: Collection<Player>) {
        players.forEach { player ->
            val radius = player.getLong(LongPropertiesEnum.SIGHT_RADIUS)!!
            updates.stream().filter { message ->
                if (message is StateChangeData) {
                    val data = message
                    return@filter player.id == data.id
                }
                if (message is AppearData) {
                    if (message.id == player.id) {
                        if (player.appeared) {
                            return@filter false
                        } else {
                            player.appeared = true
                            return@filter true
                        }
                    }
                    if (Math.abs(message.x - player.getLong(X)!!) > radius)
                        return@filter false
                    if (Math.abs(message.y - player.getLong(Y)!!) > radius)
                        return@filter false

                    // todo: think about caching
                    val key = message.x.toString() + ":" + message.y
                    if (player.cache.containsKey(key) && player.cache[key] == message)
                        return@filter false
                    else {
                        player.cache.put(key, message)
                    }
                    return@filter true
                }
                false
            }.forEach({ player.enqueueResponse(it) })
        }
    }

    companion object {

        private val LOG = LoggerFactory.getLogger(LocationWorkerManager::class.java)
    }
}
